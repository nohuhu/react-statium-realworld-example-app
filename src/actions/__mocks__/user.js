export const login = jest.fn();
export const logout = jest.fn();
export const register = jest.fn();
export const saveUserSettings = jest.fn();
