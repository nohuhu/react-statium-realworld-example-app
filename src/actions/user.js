import getApi from '../api.js';
import { displayAlert } from './alerts.js';

export const getUser = async () => {
  const token = window.localStorage.getItem('jwtToken');
  let user = null;

  if (token) {
    try {
      const api = getApi(token);
      user = await api.User.current();
    }
    catch (e) {
      // Token expired, etc. Default API is tokenless, so no need to set it again.
    }
  }

  return user;
};

export const setUser = async ({ set }, user) => {
  const token = user?.token;

  // This will return non-authenticated API if token is undefined
  const api = getApi(token);
  
  await set({ api, user });

  if (token) {
    window.localStorage.setItem('jwtToken', token);
  }
  else {
    window.localStorage.removeItem('jwtToken');
  }
};

export const login = async ({ state, set, dispatch }) => {
  await set({ busy: true });

  const { api, email, password } = state;

  try {
    const user = await api.User.login(email, password);

    // We don't reset busy flag after dispatching setUser
    // because the Login page component that dispatches this
    // action will navigate to / if state.user is defined.
    // At that point the Login component will be unmounted
    // and trying to update its store will throw an error.
    // We navigate in the Login component instead of here
    // because we want the redirect to happen upon Login
    // rendering if the user is already logged in, before
    // this action has a chance to be dispatched.
    await dispatch(setUser, user);
  }
  catch (e) {
    await set({
      busy: false,
      serverErrors: e.response?.errors ?? {
        'Login failed': ['Invalid e-mail or password'],
      },
    });
  }
};

// Reset the user object upstream. This will clean up the JWT token as well.
export const logout = async ({ data, dispatch }) => {
  await dispatch(setUser, null);

  data.navigate('/login');
};

export const register = async ({ data, state, set, dispatch }) => {
  const { api, username, email, password } = state;

  try {
    // Prevent user interaction while the API call is being made
    await set({ busy: true });

    const user = await api.User.register(username, email, password);

    // We get here only after the new user record has been created on the server
    await dispatch(setUser, user);
    await set({ busy: false });

    await dispatch(displayAlert, {
      type: "success",
      text: `Successfully created user ${username}!`,
    })

    data.navigate('/');
  }
  catch (e) {
    await set({
      serverErrors: e?.response?.errors ?? { error: `Unspecified server error: ${e}` },
      busy: false,
    });
  }
};

export const saveUserSettings = async ({ state, set, dispatch }) => {
  const { api, image, username, bio, email, password } = state;

  await set({ busy: true });

  try {
    const user = await api.User.save({
      image,
      username,
      bio,
      email,
      password,
    });

    // Update the user object upstream
    await dispatch(setUser, user);

    await set({ busy: false });

    await dispatch(displayAlert, {
      type: "success",
      text: `Settings were saved successfully!`,
    })
  }
  catch (e) {
    await set({
      serverErrors: e?.response?.errors ?? { error: `Unspecified server error: ${e}` },
      busy: false,
    });

    await dispatch(displayAlert, {
      type: "danger",
      text: `There were errors while trying to save settings...`,
    });
  }
};
