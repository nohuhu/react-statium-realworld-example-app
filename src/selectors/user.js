export const getUser = ({ state }) => state.user;

export const isCurrentUser = ({ state }) =>
  state.user && state.user?.username === state.profile?.username;

export const getUserImage = ({ state }) => state.user?.image;
export const getUserName = ({ state }) => state.user?.username;
export const getUserBio = ({ state }) => state.user?.bio;
export const getUserEmail = ({ state }) => state.user?.email;
