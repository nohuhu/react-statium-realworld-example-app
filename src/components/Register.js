import React from 'react';
import Store from 'statium';
import { Link } from 'react-router-dom';

import { register } from '../actions/user.js';
import { emailRe, inputCls, haveErrors } from '../util/form.js';

import ErrorList from './ErrorList.js';
import LoadMask from './LoadMask.js';

const initialState = {
  busy: false,
  errors: {},
  serverErrors: null,
  username: '',
  email: '',
  password: '',
  password2: '',
};

const Register = () => (
  <Store tag="Register" state={initialState} reduceState={validate}>
  {({ state, set, dispatch }) => {
    const onSubmit = e => {
      e.preventDefault();
      dispatch(register);
    };

    return (
      <div className="auth-page">
        <LoadMask loading={state.busy} />

        <div className="container page">
          <div className="row">
            <div className="col-md-6 offset-md-3 col-xs-12">
              <h1 className="text-xs-center">Sign Up</h1>

              <p className="text-xs-center">
                <Link to="/login">Have an account?</Link>
              </p>

              <ErrorList errors={state.serverErrors || {}} />

              <form onSubmit={onSubmit}>
                <fieldset>
                  <fieldset className="form-group">
                    <input type="text"
                      className={inputCls(state.errors.username)}
                      placeholder="Username"
                      value={state.username}
                      // Resetting server errors on input field change allows us to proceed
                      // after we submitted the form once and got back an error.
                      onChange={e => set({ username: e.target.value, serverErrors: null })} />
                    
                    <div className="invalid-feedback">
                      {state.errors.username}
                    </div>
                  </fieldset>

                  <fieldset className="form-group">
                    <input type="email"
                      className={inputCls(state.errors.email)}
                      placeholder="Email"
                      value={state.email}
                      onChange={e => set({ email: e.target.value, serverErrors: null })} />
                    
                    <div className="invalid-feedback">
                      {state.errors.email}
                    </div>
                  </fieldset>

                  <fieldset className="form-group">
                    <input type="password"
                      className={inputCls(state.errors.password)}
                      placeholder="Password"
                      value={state.password}
                      onChange={e => set({ password: e.target.value, serverErrors: null })} />
                    
                    <div className="invalid-feedback">
                      {state.errors.password}
                    </div>
                  </fieldset>

                  <fieldset className="form-group">
                    <input type="password"
                      className={inputCls(state.errors.password2)}
                      placeholder="Confirm password"
                      value={state.password2}
                      onChange={e => set({ password2: e.target.value, serverErrors: null })} />
                    
                    <div className="invalid-feedback">
                      {state.errors.password2}
                    </div>
                  </fieldset>
                </fieldset>

                <button type="submit" className="btn btn-lg btn-primary pull-xs-right"
                  disabled={state.busy || !isReady(state)}>
                  Sign up
                </button>
              </form>
            </div>
          </div>
        </div>
      </div>
    );
  }}
  </Store>
);

export default Register;

// This is very arbitrary, a real application likely needs
// a more sophisticated form validation approach
const validate = ({ username, email, password, password2 }) => ({
  errors: {
    username: username !== '' && username.length < 3
      ? 'User name should be longer than 3 characters'
      : null,

    email: email !== '' && !emailRe.test(email)
      ? 'E-mail address is invalid'
      : null,

    password: password !== '' && password.length < 3
      ? 'Invalid password: should be longer than 3 characters'
      : null,
    password2: password2 !== '' && password2 !== password
      ? 'Passwords do not match'
      : null,
  },
});

const isReady = ({ username, email, password, password2, ...state }) =>
  !haveErrors(state) && username && email && password && password2;
