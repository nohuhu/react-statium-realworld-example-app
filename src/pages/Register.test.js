import React from 'react';
import { mount } from 'test/enzyme.js';
import Store from 'statium';

jest.mock('../actions/user.js');

import { register } from '../actions/user.js';

import Register from './Register.jsx';

describe("Register form", () => {
  describe("UI behavior", () => {
    it("should render", () => {
      const tree = mount(<Register />);

      expect(tree.find('Register')).toMatchSnapshot();
    });

    it("should disable submit button by default", () => {
      const tree = mount(<Register />);

      expect(tree.find('button[type="submit"]').prop('disabled')).toBe(true);
    });

    it("should enable submit button when filled in and valid", async () => {
      const tree = mount(<Register />);

      await tree.find('Store').instance().set({
        username: 'foobaroo',
        email: 'foo@bar.baz',
        password: 'foobaroo',
        password2: 'foobaroo',
      });

      tree.update();

      expect(tree.find('Register')).toMatchSnapshot();
      expect(tree.find('button[type="submit"]').prop('disabled')).toBe(false);
    });

    it("should disable submit button when there are server errors", async () => {
      const tree = mount(<Register />);

      await tree.find('Store').instance().set({
        username: 'bazuroo',
        email: 'baz@bar.baz',
        password: 'blergobonz',
        password2: 'blergobonz',
        serverErrors: {
          "email": ["has already been taken"],
          "username": ["has already been taken"],
        },
      });

      tree.update();

      const { errors } = tree.find('Store').instance().store.state;

      // No form errors are expected, all values are valid
      expect(errors).toEqual({
        username: null,
        email: null,
        password: null,
        password2: null,
      });

      // Server errors are displayed as a list above the form
      expect(tree.find('ErrorList')).toMatchSnapshot();

      expect(tree.find('button[type="submit"]').prop('disabled')).toBe(true);
    });
  });

  describe("user interaction", () => {
    beforeEach(() => {
      register.mockClear();
    });

    it.only("should dispatch register action when form is submitted", async () => {
      // state.api.User.register.mockImplementation((username, email) => ({
      //   username,
      //   email,
      //   bio: null,
      //   image: 'https://api.realworld.io/images/smiley-cyrus.jpeg',
      //   token: 'auth-token',
      // }));

      const tree = mount(
        // <Notifications>
        //   <Store tag="test" data={data} state={state}>
            <Register />
        //   </Store>
        // </Notifications>
      );

      // We have more than one Store in the tree in this test, need to specify
      // which one to get the instance for. The "tag" prop is ideal for this.
      await tree.find('Store[tag="Register"]').instance().set({
        username: 'frobbe',
        email: 'frobbe@bar.baz',
        password: 'throbbe',
        password2: 'throbbe',
      });

      tree.update();
      
      // The usual option would be to simulate the click event on the submit button.
      // This approach is better since submitting a form can be done by pressing Enter key
      // in a field as well and this way we test that, too.
      tree.find('form').simulate('submit');

      // Give the event handler a chance to fire. We cannot await for a Promise here
      // since it is not returned from the event simulation method. This is a limitation
      // of Enzyme API.
      await sleep(10);

      expect()
      // expect(state.api.User.register).toHaveBeenCalledWith('frobbe', 'frobbe@bar.baz', 'throbbe');

      // register action will change the URL upon successful user creation
      // expect(data.navigate).toHaveBeenCalledWith('/');
    });
  });
});
