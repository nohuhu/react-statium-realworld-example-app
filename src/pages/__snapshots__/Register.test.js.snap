// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register form UI behavior should disable submit button when there are server errors 1`] = `
<ErrorList
  errors={
    Object {
      "email": Array [
        "has already been taken",
      ],
      "username": Array [
        "has already been taken",
      ],
    }
  }
>
  <ul
    className="error-messages"
  >
    <li
      key="email"
    >
      email: 
      has already been taken
    </li>
    <li
      key="username"
    >
      username: 
      has already been taken
    </li>
  </ul>
</ErrorList>
`;

exports[`Register form UI behavior should enable submit button when filled in and valid 1`] = `
<Register>
  <Store
    data={Object {}}
    reduceState={[Function]}
    state={
      Object {
        "busy": false,
        "email": "",
        "errors": Object {},
        "password": "",
        "password2": "",
        "serverErrors": null,
        "username": "",
      }
    }
    tag="Register"
  >
    <div
      className="auth-page"
    >
      <LoadMask
        loading={false}
      >
        <div
          className="load-mask hidden"
        >
          <div
            className="load-mask-inner"
          >
            <i
              className="ion ion-load-c load-spinner"
              style={
                Object {
                  "fontSize": "50px",
                }
              }
            />
          </div>
        </div>
      </LoadMask>
      <div
        className="container page"
      >
        <div
          className="row"
        >
          <div
            className="col-md-6 offset-md-3 col-xs-12"
          >
            <h1
              className="text-xs-center"
            >
              Sign Up
            </h1>
            <p
              className="text-xs-center"
            >
              <Link
                to="/login"
              >
                <a
                  href="/login"
                  onClick={[Function]}
                >
                  Have an account?
                </a>
              </Link>
            </p>
            <ErrorList
              errors={Object {}}
            >
              <ul
                className="error-messages"
              />
            </ErrorList>
            <form
              onSubmit={[Function]}
            >
              <fieldset>
                <fieldset
                  className="form-group"
                >
                  <input
                    className="form-control form-control-lg is-valid"
                    onChange={[Function]}
                    placeholder="Username"
                    type="text"
                    value="foobaroo"
                  />
                  <div
                    className="invalid-feedback"
                  />
                </fieldset>
                <fieldset
                  className="form-group"
                >
                  <input
                    className="form-control form-control-lg is-valid"
                    onChange={[Function]}
                    placeholder="Email"
                    type="email"
                    value="foo@bar.baz"
                  />
                  <div
                    className="invalid-feedback"
                  />
                </fieldset>
                <fieldset
                  className="form-group"
                >
                  <input
                    className="form-control form-control-lg is-valid"
                    onChange={[Function]}
                    placeholder="Password"
                    type="password"
                    value="foobaroo"
                  />
                  <div
                    className="invalid-feedback"
                  />
                </fieldset>
                <fieldset
                  className="form-group"
                >
                  <input
                    className="form-control form-control-lg is-valid"
                    onChange={[Function]}
                    placeholder="Confirm password"
                    type="password"
                    value="foobaroo"
                  />
                  <div
                    className="invalid-feedback"
                  />
                </fieldset>
              </fieldset>
              <button
                className="btn btn-lg btn-primary pull-xs-right"
                disabled={false}
                type="submit"
              >
                Sign up
              </button>
            </form>
          </div>
        </div>
      </div>
    </div>
  </Store>
</Register>
`;

exports[`Register form UI behavior should render 1`] = `
<Register>
  <Store
    data={Object {}}
    reduceState={[Function]}
    state={
      Object {
        "busy": false,
        "email": "",
        "errors": Object {},
        "password": "",
        "password2": "",
        "serverErrors": null,
        "username": "",
      }
    }
    tag="Register"
  >
    <div
      className="auth-page"
    >
      <LoadMask
        loading={false}
      >
        <div
          className="load-mask hidden"
        >
          <div
            className="load-mask-inner"
          >
            <i
              className="ion ion-load-c load-spinner"
              style={
                Object {
                  "fontSize": "50px",
                }
              }
            />
          </div>
        </div>
      </LoadMask>
      <div
        className="container page"
      >
        <div
          className="row"
        >
          <div
            className="col-md-6 offset-md-3 col-xs-12"
          >
            <h1
              className="text-xs-center"
            >
              Sign Up
            </h1>
            <p
              className="text-xs-center"
            >
              <Link
                to="/login"
              >
                <a
                  href="/login"
                  onClick={[Function]}
                >
                  Have an account?
                </a>
              </Link>
            </p>
            <ErrorList
              errors={Object {}}
            >
              <ul
                className="error-messages"
              />
            </ErrorList>
            <form
              onSubmit={[Function]}
            >
              <fieldset>
                <fieldset
                  className="form-group"
                >
                  <input
                    className="form-control form-control-lg is-valid"
                    onChange={[Function]}
                    placeholder="Username"
                    type="text"
                    value=""
                  />
                  <div
                    className="invalid-feedback"
                  />
                </fieldset>
                <fieldset
                  className="form-group"
                >
                  <input
                    className="form-control form-control-lg is-valid"
                    onChange={[Function]}
                    placeholder="Email"
                    type="email"
                    value=""
                  />
                  <div
                    className="invalid-feedback"
                  />
                </fieldset>
                <fieldset
                  className="form-group"
                >
                  <input
                    className="form-control form-control-lg is-valid"
                    onChange={[Function]}
                    placeholder="Password"
                    type="password"
                    value=""
                  />
                  <div
                    className="invalid-feedback"
                  />
                </fieldset>
                <fieldset
                  className="form-group"
                >
                  <input
                    className="form-control form-control-lg is-valid"
                    onChange={[Function]}
                    placeholder="Confirm password"
                    type="password"
                    value=""
                  />
                  <div
                    className="invalid-feedback"
                  />
                </fieldset>
              </fieldset>
              <button
                className="btn btn-lg btn-primary pull-xs-right"
                disabled={true}
                type="submit"
              >
                Sign up
              </button>
            </form>
          </div>
        </div>
      </div>
    </div>
  </Store>
</Register>
`;
